### ISSUE TITLE
Add support for 'redirect' and 'fixed-response' into lb_listener_rule action type

### ISSUE THREAD

**Issue by julianxhokaxhiu** [terraform-providers/terraform-provider-aws#5344]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

AWS added support for ELB Redirects, which is very useful to do an out-of-the-box experience for HTTP to HTTPS redirects without using an EC2 instance with an HTTP server listening in there.

It would be cool if you could add the support for this feature.

### New or Affected Resource(s)

* aws_lb_listener_rule

### Potential Terraform Configuration

```hcl
resource "aws_lb_listener_rule" "static" {
  listener_arn = "${aws_lb_listener.front_end.arn}"
  priority     = 100

  action {
    type = "redirect" # "fixed-response"
    # the rest has to be concepted
  }
}
```

### References

Official announcement: https://aws.amazon.com/about-aws/whats-new/2018/07/elastic-load-balancing-announces-support-for-redirects-and-fixed-responses-for-application-load-balancer/

Documentation:
* https://docs.aws.amazon.com/elasticloadbalancing/latest/application/load-balancer-listeners.html#redirect-actions
* https://docs.aws.amazon.com/elasticloadbalancing/latest/application/load-balancer-listeners.html#fixed-response-actions

**Comment by jianyuan** [comment]
**I'm interested to work on this feature.**

The `aws_lb_listener` resource would be affected as well.

As for the potential Terraform configuration, I propose the following:

#### Redirect

```hcl
resource "aws_lb_listener" "web_port80" {
  load_balancer_arn = "${aws_lb.web_alb.arn}"
  port              = 80
  protocol          = "HTTP"

  default_action {
    type = "redirect"
    host = "#{host}"
    path = "/#{path}"
    port = "#{port}"
    protocol = "#{protocol}"
    query = "#{query}"
    status_code = "301"
  }
}
```

#### Fixed response
```hcl
resource "aws_lb_listener" "web_port80" {
  load_balancer_arn = "${aws_lb.web_alb.arn}"
  port              = 80
  protocol          = "HTTP"

  default_action {
    type = "fixed-response"
    content_type = "text/plain"
    message_body = "Hello"
    status_code = "200"
  }
}
```


Config for `aws_lb_listener_rule` would be similar.

**Comment by bflad** [comment]
Caution: setting these types of rules outside Terraform will cause the AWS provider to crash in all versions up through the latest (version 1.29.0 as of this writing): #5357

I will be submitting a pull request today to prevent this type of crash across all ALB/NLB resources and we should probably wait on those fixes before adding this feature.

**Comment by bflad** [comment]
Crash prevention pull request submitted: #5367

**Comment by bflad** [comment]
The crash fix has been merged into master so it should be okay to work with the code for this on top of those changes. 👍

**Comment by jianyuan** [comment]
We now have 3 different types of actions: "fixed-response", "forward" and "redirect", each with different sets of parameters.

Would it be better if the action config for selected action type is grouped in a block instead of flattening them at the top level?

Examples:

```
resource "aws_lb_listener_rule" "static" {
  action {
    type = "forward"
    forward {
      target_group_arn = "..."
    }
  }
}
```

```
resource "aws_lb_listener_rule" "static" {
  action {
    type = "fixed-response"
    fixed_response {
      content_type = "text/plain"
      message_body = "Hello"
      status_code = "200"
    }
  }
}
```

```
resource "aws_lb_listener_rule" "static" {
  action {
    type = "redirect"
    redirect {
	  host = "#{host}"
	  path = "/#{path}"
	  port = "#{port}"
	  protocol = "#{protocol}"
	  query = "#{query}"
	  status_code = "HTTP_301"
    }
  }
}
```

The Action datatype now groups action type-specific config in a nested object, so doing it this way would be (almost) a 1:1 mapping: https://docs.aws.amazon.com/elasticloadbalancing/latest/APIReference/API_Action.html

Also, I noticed that the status_code for redirect action is prefixed by `HTTP_` whereas that for fixed-response action is not:
https://docs.aws.amazon.com/elasticloadbalancing/latest/APIReference/API_RedirectActionConfig.html (Valid Values: HTTP_301 | HTTP_302)
https://docs.aws.amazon.com/elasticloadbalancing/latest/APIReference/API_FixedResponseActionConfig.html (Pattern: ^(2|4|5)\d\d$)

**Comment by bflad** [comment]
We recommend following the SDK structure in most cases, so separate blocks in this case would be appropriate. This allows us to easily add the argument validation as well as keep the code simpler.

**Comment by bacoboy** [comment]
Since the scope of this includes `fixed-response` can we update the title?  I was interested in that specifically and took me a bit to find this issue and already being worked on.  Thx.

Additionally, I didn't see anything about adding headers in the API links above so not clear how I'd set something like an [`HSTS` header](https://en.wikipedia.org/wiki/HTTP_Strict_Transport_Security) in a response -- because the ALB in forward mode leaves that to the backing app.

**Comment by sklemmer** [comment]
Are there any plans to add authentication too?

**Comment by bflad** [comment]
@sklemmer feature requests for authentication can be found at #4707 and #4711 😄 

I'm hoping to review the open pull request for `fixed-response`/`redirect` (#5430) sometime this week.

**Comment by bflad** [comment]
Support for `fixed-response` and `redirect` actions has been merged into master via #5430 and will release with version 1.33.0 of the AWS provider, later this week. 🚀

**Comment by bflad** [comment]
This has been released in [version 1.33.0 of the AWS provider](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md#1330-august-22-2018). Please see the [Terraform documentation on provider versioning](https://www.terraform.io/docs/configuration/providers.html#provider-versions) or reach out if you need any assistance upgrading.
